client/consensus/instant-seal/src/lib.rs

```diff
@@ -0,0 +1,183 @@
+
+use consensus_common::{
+	self, BlockImport, Environment, Proposer, BlockCheckParams,
+	ForkChoiceStrategy, BlockImportParams, BlockOrigin,
+	ImportResult, SelectChain,
+};
+use consensus_common::import_queue::{BasicQueue, CacheKeyId, Verifier, BoxBlockImport};
+use sr_primitives::traits::{Block as BlockT};
+use sr_primitives::Justification;
+use parking_lot::Mutex;
+use futures::prelude::*;
+use transaction_pool::txpool::{self, Pool as TransactionPool};
+
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::time::Duration;
+
+/// The synchronous block-import worker of the engine.
+pub struct InstantSealBlockImport<I> {
+	inner: I,
+}
+
+impl<I> From<I> for InstantSealBlockImport<I> {
+	fn from(i: I) -> Self {
+		InstantSealBlockImport { inner: i }
+	}
+}
+
+impl<B: BlockT, I: BlockImport<B>> BlockImport<B> for InstantSealBlockImport<I> {
+	type Error = I::Error;
+
+	fn check_block(&mut self, block: BlockCheckParams<B>)
+	               -> Result<ImportResult, Self::Error>
+	{
+		self.inner.check_block(block)
+	}
+
+	fn import_block(
+		&mut self,
+		block: BlockImportParams<B>,
+		cache: HashMap<CacheKeyId, Vec<u8>>,
+	) -> Result<ImportResult, Self::Error> {
+		// TODO: strip out post-digest.
+
+		self.inner.import_block(block, cache)
+	}
+}
+
+/// The verifier for the instant seal engine; instantly finalizes.
+struct InstantSealVerifier;
+
+impl<B: BlockT> Verifier<B> for InstantSealVerifier {
+	fn verify(
+		&mut self,
+		origin: BlockOrigin,
+		header: B::Header,
+		justification: Option<Justification>,
+		body: Option<Vec<B::Extrinsic>>,
+	) -> Result<(BlockImportParams<B>, Option<Vec<(CacheKeyId, Vec<u8>)>>), String> {
+		let import_params = BlockImportParams {
+			origin,
+			header,
+			justification,
+			post_digests: Vec::new(),
+			body,
+			finalized: true,
+			auxiliary: Vec::new(),
+			fork_choice: ForkChoiceStrategy::LongestChain,
+			allow_missing_state: false,
+		};
+
+		Ok((import_params, None))
+	}
+}
+
+/// Instantiate the import queue for the instant-seal consensus engine.
+pub fn import_queue<B: BlockT>(block_import: BoxBlockImport<B>) -> BasicQueue<B>
+{
+	BasicQueue::new(
+		InstantSealVerifier,
+		block_import,
+		None,
+		None,
+	)
+}
+
+/// Creates the background authorship task for the instant seal engine.
+pub async fn run_instant_seal<B, E, A, C>(
+	block_import: BoxBlockImport<B>,
+	env: E,
+	pool: Arc<TransactionPool<A>>,
+	select_chain: C,
+	inherent_data_providers: inherents::InherentDataProviders,
+) -> ()
+	where
+		B: BlockT + 'static,
+		E: Environment<B> + 'static,
+		A: txpool::ChainApi + 'static,
+		C: SelectChain<B> + 'static,
+{
+	let block_import = Arc::new(Mutex::new(block_import));
+	let env = Arc::new(Mutex::new(env));
+	let select_chain = Arc::new(select_chain);
+	let inherent_data_providers = Arc::new(inherent_data_providers);
+
+	// propose a new block everytime a transaction is imported
+	pool.import_notification_stream()
```

Comment:
Wouldn't that lead to having two blocks proposed in parallel at the same height? Since the code is asynchronous here, we might be spawn two futures by accident.

I think it would be best to make sure blocks are created synchronously only in one `future` and the notification stream just triggers the other task to do the work.

Reply:
Great catch, yeah I was worried about that initially, but I checked the futures crate to see how it handled `for_each` and it appears that it will poll the future produced by the closure to completion before polling the next item of the stream. 

https://github.com/rust-lang-nursery/futures-rs/blob/c4291883bcc9d3ae8cc3186b2eec27b096dafb66/futures-util/src/stream/stream/for_each.rs#L70-L87

there's also `for_each_concurrent` which will do exactly what you've described.

Reply:
Cool. Good point, didn't know about `for_each_concurrent` actually. Anyways, if that's the case we should put a guard inside the closure and never produce a block if there is no pending transactions.
Currently what may happen is:
1. You have 1000 transactions that get imported to the pool concurrently.
2. This means you will get 1000 notifications on the `import_notifications_stream`.
3. Which in turn means that you will produce 1000 blocks
4. But the first block will most likely contain most of the transactions (all transactions that were able to be imported already)
5. Then second block will have the rest
6. And the rest 998 blocks will be empty.


Also it would be good to write some tests for this engine, to make sure we don't run into similar issues as in Parity Ethereum.